### PR TITLE
KL Divergence Metrics

### PR DESCRIPTION
**Pull request by andrewztan**
## What do these changes do?

Added KL divergence between the logits of the workers vs logits of the learner (current policy). Collects KL divergence for every step in each iteration in an array. Compute mean, median, max, min KL values for each iteration.

### CODE REVIEW

**Comment by andrewztan** [comment]
@ericl @richardliaw

File: python/ray/rllib/agents/impala/vtrace_policy_graph.py [hunk 1]
```diff
@@ -225,6 +235,11 @@ def to_batches(tensor):
                 "vf_explained_var": explained_variance(
                     tf.reshape(self.loss.vtrace_returns.vs, [-1]),
                     tf.reshape(to_batches(values)[:-1], [-1])),
+                "KLs": self.KLs,
```
Comment (ericl) [inline python/ray/rllib/agents/impala/vtrace_policy_graph.py]: Let's drop self.KLs
```suggestion
```

File: python/ray/rllib/agents/impala/vtrace_policy_graph.py [hunk 2]
```diff
@@ -184,6 +185,15 @@ def to_batches(tensor):
             clip_rho_threshold=self.config["vtrace_clip_rho_threshold"],
             clip_pg_rho_threshold=self.config["vtrace_clip_pg_rho_threshold"])
 
+        # Compute KL divergence metrics
+        model_dist = Categorical(self.model.outputs)
+        behaviour_dist = Categorical(behaviour_logits)
+        self.KLs = model_dist.kl(behaviour_dist)
```
Comment (ericl) [inline python/ray/rllib/agents/impala/vtrace_policy_graph.py]:
```suggestion
```

File: python/ray/rllib/agents/impala/vtrace_policy_graph.py [hunk 3]
```diff
@@ -184,6 +185,15 @@ def to_batches(tensor):
             clip_rho_threshold=self.config["vtrace_clip_rho_threshold"],
             clip_pg_rho_threshold=self.config["vtrace_clip_pg_rho_threshold"])
 
+        # Compute KL divergence metrics
```
Comment (ericl) [inline python/ray/rllib/agents/impala/vtrace_policy_graph.py]:
```suggestion
        # KL divergence between worker and learner logits
```

File: python/ray/rllib/agents/impala/vtrace_policy_graph.py [hunk 4]
```diff
@@ -184,6 +185,15 @@ def to_batches(tensor):
...
+        model_dist = Categorical(self.model.outputs)
+        behaviour_dist = Categorical(behaviour_logits)
+        self.KLs = model_dist.kl(behaviour_dist)
+        self.mean_KL = tf.reduce_mean(self.KLs)
+        self.max_KL = tf.reduce_max(self.KLs)
+        self.min_KL = tf.reduce_min(self.KLs)
+        self.median_KL = tf.contrib.distributions.percentile(self.KLs, 50.0)
```
Comment (ericl) [inline python/ray/rllib/agents/impala/vtrace_policy_graph.py]: How about we drop min and median as well? Min is probably not very interesting and median is ~= mean.

File: python/ray/tune/logger.py [hunk 5]
```diff
@@ -188,8 +188,20 @@ def default(self, value):
             return str(value)  # give up, just stringify it (ok for logs)
 
 
+def _reduce_mean_result_dict(result):
```
Comment (ericl) [inline python/ray/tune/logger.py]: Let's remove this as well.

**Review by ericl — COMMENTED** [review]
Looks good. Btw is it easy to add the timestamp differences as well here?

File: python/ray/rllib/agents/impala/vtrace_policy_graph.py [hunk 6]
```diff
@@ -225,6 +235,11 @@ def to_batches(tensor):
                 "vf_explained_var": explained_variance(
                     tf.reshape(self.loss.vtrace_returns.vs, [-1]),
                     tf.reshape(to_batches(values)[:-1], [-1])),
+                "KLs": self.KLs,
+                "mean_KL": self.mean_KL,
+                "min_KL": self.min_KL,
```
Comment (ericl) [inline python/ray/rllib/agents/impala/vtrace_policy_graph.py]:
```suggestion
                "min_KL": self.min_KL,
```

File: python/ray/rllib/agents/impala/vtrace_policy_graph.py [hunk 7]
```diff
@@ -225,6 +235,11 @@ def to_batches(tensor):
...
+                "max_KL": self.max_KL,
+                "median_KL": self.median_KL,
```
Comment (ericl) [inline python/ray/rllib/agents/impala/vtrace_policy_graph.py]:
```suggestion
                "median_KL": self.median_KL,
```

File: python/ray/rllib/agents/impala/vtrace_policy_graph.py [hunk 8]
```diff
@@ -184,6 +185,15 @@ def to_batches(tensor):
             clip_rho_threshold=self.config["vtrace_clip_rho_threshold"],
             clip_pg_rho_threshold=self.config["vtrace_clip_pg_rho_threshold"])
 
+        # Compute KL divergence metrics
+        model_dist = Categorical(self.model.outputs)
+        behaviour_dist = Categorical(behaviour_logits)
+        self.KLs = model_dist.kl(behaviour_dist)
+        self.mean_KL = tf.reduce_mean(self.KLs)
+        self.max_KL = tf.reduce_max(self.KLs)
+        self.min_KL = tf.reduce_min(self.KLs)
```
Comment (ericl) [inline python/ray/rllib/agents/impala/vtrace_policy_graph.py]:
```suggestion
```

File: python/ray/tune/logger.py [hunk 9]
```diff
@@ -188,8 +188,20 @@ def default(self, value):
...
+        return np.mean(result)
+    elif not isinstance(result, dict):
+        return result
+    else:
+        return {key: _reduce_mean_result_dict(val) for key, val in result.items()}
+
+
 def pretty_print(result):
-    result = result.copy()
+    result = _reduce_mean_result_dict(result.copy())
```
Comment (ericl) [inline python/ray/tune/logger.py]:
```suggestion
    result = result.copy()
```

File: python/ray/rllib/agents/impala/vtrace_policy_graph.py (hunk 1 above)
Comment (andrewztan) [reply python/ray/rllib/agents/impala/vtrace_policy_graph.py]: self.KLs was used to create the histograms on tensorboard. is that something we're still interested in?

**Review by ericl — APPROVED** [review]
